File: app/src/main/java/com/myflightbook/android/WebServices/CommitFlightSvc.java
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android.WebServices;

import com.myflightbook.android.MFBMain;
import com.myflightbook.android.Marshal.MarshalDate;
import com.myflightbook.android.Marshal.MarshalDouble;
import com.myflightbook.android.R;

import org.ksoap2.serialization.PropertyInfo;
import org.ksoap2.serialization.SoapObject;
import org.ksoap2.serialization.SoapSerializationEnvelope;

import java.util.Date;

import Model.FlightProperty;
import Model.LatLong;
import Model.LogbookEntry;
import Model.MFBImageInfo;
import Model.MFBUtil;
import Model.PostingOptions;

public class CommitFlightSvc extends MFBSoap {

	public CommitFlightSvc()
	{
	}

	@Override
	public void AddMappings(SoapSerializationEnvelope e)
	{
		e.addMapping(NAMESPACE, "CommitFlightWithOptionsResult", LogbookEntry.class);
		e.addMapping(NAMESPACE, "le", LogbookEntry.class);
		e.addMapping(NAMESPACE, "PostingOptions", PostingOptions.class);
		e.addMapping(NAMESPACE, "LogbookEntry", LogbookEntry.class);
		e.addMapping(NAMESPACE, "CustomFlightProperty", FlightProperty.class);
		e.addMapping(NAMESPACE, "MFBImageInfo", MFBImageInfo.class);
		e.addMapping(NAMESPACE, "LatLong", LatLong.class);
		
		MarshalDate mdt = new MarshalDate();
		MarshalDouble md = new MarshalDouble();
		mdt.register(e);
		md.register(e);
	}
	
	public Boolean FCommitFlight(String szAuthToken, LogbookEntry le, PostingOptions po)
	{
		SoapObject Request = setMethod("CommitFlightWithOptions");
    	Request.addProperty("szAuthUserToken", szAuthToken);
    	PropertyInfo piLe = new PropertyInfo();
    	PropertyInfo pipo = new PropertyInfo();
    	piLe.name = "le";
    	piLe.type = "LogbookEntry";
    	piLe.setValue(le);
    	piLe.namespace = NAMESPACE;
    	pipo.name = "po";
    	pipo.type = "PostingOptions";
    	pipo.setValue(po);
    	pipo.namespace = NAMESPACE;
    	Request.addProperty(piLe);
    	Request.addProperty(pipo);
    	
    	// The date of flight is done in local time; need to convert it to
    	// a UTC time that looks like the correct local time so that it records
    	// correctly over the wire (same as in iPhone version.
    	
    	// save the date, since we're making a live copy
    	Date dtSave = le.dtFlight;
    	
    	le.dtFlight = MFBUtil.UTCDateFromLocalDate(le.dtFlight);
    	
    	if (le.rgFlightImages == null)
    		le.getImagesForFlight();
    	
    	MFBImageInfo[] rgmfbii = le.rgFlightImages;    		
    	assert(rgmfbii != null); // may have zero length, but that's ok
    	
    	le.rgFlightImages = new MFBImageInfo[0];
    	SoapObject result = (SoapObject) Invoke();
		
		if (result == null)
		{
			le.rgFlightImages = rgmfbii; // restore the images.
			setLastError("Failed to save flight - " + getLastError());
		}
		else
		{
			try
			{
				LogbookEntry leReturn = new LogbookEntry();
				leReturn.FromProperties(result);
				le.szError = leReturn.szError;
				
				int iImg = 1;
				for (MFBImageInfo mfbii : rgmfbii)
				{
	    			String szFmtUploadProgress = MFBMain.GetMainContext().getString(R.string.prgUploadingImages);
					String szStatus = String.format(szFmtUploadProgress, iImg, rgmfbii.length);
					if (m_Progress != null)
						m_Progress.NotifyProgress((iImg * 100) / rgmfbii.length, szStatus);
					// skip anything that is already on the server.
					if (!mfbii.IsOnServer())
					{
						mfbii.setKey(leReturn.idFlight);
						mfbii.UploadPendingImage(mfbii.getID());
					}
					iImg++;
				}
				le.DeletePendingImagesForFlight();
			}
			catch (Exception e)
			{
				setLastError(getLastError() + e.getMessage());
			}
		}
		
		le.dtFlight = dtSave;
		
		le.szError = getLastError();
    	return getLastError().length() == 0;
	}	
}

File: app/src/main/java/com/myflightbook/android/webservices/CommitFlightSvc.kt
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2025 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android.webservices

import android.content.Context
import android.util.Log
import com.myflightbook.android.R
import com.myflightbook.android.marshal.MarshalDate
import com.myflightbook.android.marshal.MarshalDouble
import model.*
import org.ksoap2.serialization.PropertyInfo
import org.ksoap2.serialization.SoapObject
import org.ksoap2.serialization.SoapSerializationEnvelope

class CommitFlightSvc : MFBSoap() {
    override fun addMappings(e: SoapSerializationEnvelope) {
        e.addMapping(NAMESPACE, "CommitFlightWithOptionsResult", LogbookEntry::class.java)
        e.addMapping(NAMESPACE, "le", LogbookEntry::class.java)
        e.addMapping(NAMESPACE, "LogbookEntry", LogbookEntry::class.java)
        e.addMapping(NAMESPACE, "CustomFlightProperty", FlightProperty::class.java)
        e.addMapping(NAMESPACE, "MFBImageInfo", MFBImageInfo::class.java)
        e.addMapping(NAMESPACE, "LatLong", LatLong::class.java)
        val mdt = MarshalDate()
        val md = MarshalDouble()
        mdt.register(e)
        md.register(e)
    }

    fun checkFlight(szAuthToken: String?, le: LogbookEntry, c: Context) : Array<String> {
        val request = setMethod("CheckFlight")
        request.addProperty("szAuthUserToken", szAuthToken)
        val piLe = PropertyInfo()
        piLe.name = "le"
        piLe.type = "LogbookEntry"
        piLe.value = le
        piLe.namespace = NAMESPACE
        request.addProperty(piLe)

        val result : ArrayList<String> = ArrayList()
        val r = invoke(c) as SoapObject?
        if (r != null) {
            try {
                for (i in 0 until r.propertyCount)
                    result.add(r.getProperty(i).toString())
            } catch (e: Exception) {
                lastError += e.message
            }
        }
        return result.toTypedArray()
    }

    fun fCommitFlightForUser(szAuthToken: String?, le: LogbookEntry, c: Context): Boolean {
        val request = setMethod("CommitFlightWithOptions")
        request.addProperty("szAuthUserToken", szAuthToken)
        val piLe = PropertyInfo()
        val pipo = PropertyInfo()
        piLe.name = "le"
        piLe.type = "LogbookEntry"
        piLe.value = le
        piLe.namespace = NAMESPACE
        pipo.name = "po"
        pipo.type = "PostingOptions"
        pipo.value = null
        pipo.namespace = NAMESPACE
        request.addProperty(piLe)
        request.addProperty(pipo)

        if (le.rgFlightImages == null) le.imagesForFlight
        val rgmfbii = le.rgFlightImages
            ?: throw NullPointerException("le.rgFlightImages is null")
        le.rgFlightImages = arrayOf()
        val result = invoke(c) as SoapObject? ?: return false
        try {
            val leReturn = LogbookEntry()
            leReturn.fromProperties(result)
            le.szError = leReturn.szError
            le.idFlight = leReturn.idFlight
            var iImg = 1
            for (mfbii in rgmfbii) {
                val szFmtUploadProgress = c.getString(R.string.prgUploadingImages)
                val szStatus = String.format(szFmtUploadProgress, iImg, rgmfbii.size)
                mProgress?.notifyProgress(
                    iImg * 100 / rgmfbii.size,
                    szStatus
                )
                // skip anything that is already on the server.
                if (!mfbii.isOnServer()) {
                    mfbii.setKey(leReturn.idFlight)
                    if (!mfbii.uploadPendingImage(mfbii.id, c)) Log.w(
                        MFBConstants.LOG_TAG,
                        "Image Upload failed"
                    )
                }
                iImg++
            }
            le.deletePendingImagesForFlight()
        } catch (e: Exception) {
            lastError += e.message
        }
        le.szError = lastError
        return lastError.isEmpty()
    }
}